src/ucs/type/float.h

```diff
@@ -0,0 +1,232 @@
+/**
+* Copyright (C) Mellanox Technologies Ltd. 2021.  ALL RIGHTS RESERVED.
+*
+* See file LICENSE for terms.
+*/
+
+#ifndef UCS_TYPE_FLOAT_H
+#define UCS_TYPE_FLOAT_H
+
+#include <ieee754.h>
+
+#include <ucs/sys/preprocessor.h>
+#include <ucs/debug/assert.h>
+#include <ucs/arch/bitops.h>
+
+BEGIN_C_DECLS
+
+
+typedef uint8_t ucs_fp8_t;
+
+
+/**
+ * Bits number in the exponent part of a packed floating-point number
+ */
+#define _UCS_FP8_EXPONENT_BITS (4)
+
+
+/**
+ * Bits number in the mantissa part of a packed floating-point number
+ */
+#define _UCS_FP8_MANTISSA_BITS (4)
+
+
+/**
+ * The ration of the value obtained after packing and unpacking to
+ * the original number */
+#define UCS_FP8_PRECISION \
+    ((float)UCS_MASK(_UCS_FP8_MANTISSA_BITS) / UCS_BIT(_UCS_FP8_MANTISSA_BITS))
+
+
+/**
+ * Bits number in the exponent part of an IEEE754 double
+ */
+#define _UCS_FP8_IEEE_EXPONENT_BITS (11)
+
+
+/**
+ * Bits number in the significant mantissa part of an IEEE754 double
+ */
+#define _UCS_FP8_IEEE_MANTISSA_BITS (20)
+
+
+/**
+ * A special value of exponent which represents NaN in an IEEE754 double
+ */
+#define _UCS_FP8_IEEE_NAN_EXPONENT (UCS_BIT(_UCS_FP8_IEEE_EXPONENT_BITS) - 1)
+
+
+/**
+ * A special value of exponent which represents NaN in a packed floating-point number
+ */
+#define _UCS_FP8_NAN (UCS_MASK(_UCS_FP8_EXPONENT_BITS))
+
+
+/**
+ * The offset of an IEEE754 exponent representation from a packed exponent representation
+ */
+#define _UCS_FP8_EXPONENT_OFFSET (IEEE754_DOUBLE_BIAS - 1)
+
+
+/**
+ * Internal macro to construct floating-point type identifier from a name and a suffix
+ */
+#define _UCS_FP8_IDENTIFIER(_name, _suffix) \
+    UCS_PP_TOKENPASTE3(ucs_fp8_, _name, _suffix)
+
+
+/**
+ * Mask the exponent part of a packed floating-point number
+ */
+#define _UCS_FP8_EXPONENT_MASK (UCS_MASK(_UCS_FP8_EXPONENT_BITS))
+
+
+/**
+ * pack a double-precision floating-point number in a given range to a single byte.
+ * The packing is lossy and the unpacked number is assumed to be
+ * non-negative.
+ * 
+ * @param min   Min supported value (assumed to be a power of 2)
+ * @param max   Max supported value (assumed to be a power of 2)
+ * @param value Pack this number
+ * 
+ * @return A single byte which represents the given number
+ */
+static UCS_F_ALWAYS_INLINE ucs_fp8_t ucs_fp8_pack(uint64_t min, uint64_t max,
```

Comment:
but for latency the min/max could be non-integer..

Reply:
I assumed it would be an integer (see comment below on latency). WIll adjust.